feat(cli): Allow adding admin password using docker secrets (#1199)

nenadilic84 opened the pull request:
Close #1199

deviantony commented:
Hey @nenadilic84 

Thanks for the contribution ! I just had a quick look at your PR, it seems that you expect the file to contain the encrypted password for the administrator user.

The whole point of using secrets is to expose the password unencrypted in a file.

The goal is to do something like this:

```
echo $PORTAINER_PASS | docker secret create portainer_password -
docker service create --name portainer \
  --secret portainer_password \
  -p 9000:9000 \
  --mount type=bind,source=/var/run/docker.sock,target=/var/run/docker.sock \
  portainer/portainer \
  --admin-password-file '/run/secrets/portainer-password'
```

As you can see here, `$PORTAINER_PASS` is the unencrypted password. I don't think that's necessary to encrypt it in the secret & in the file.

nenadilic84 commented:
Got it, will change that! Thanks for the feedback

nenadilic84 commented:
Hey @deviantony, let me know if it's this what you had in mind.

deviantony commented on api/cli/cli.go:
```diff
@@ -45,6 +46,7 @@ func (*Service) ParseFlags(version string) (*portainer.CLIFlags, error) {
 		SSLCert:           kingpin.Flag("sslcert", "Path to the SSL certificate used to secure the Portainer instance").Default(defaultSSLCertPath).String(),
 		SSLKey:            kingpin.Flag("sslkey", "Path to the SSL key used to secure the Portainer instance").Default(defaultSSLKeyPath).String(),
 		AdminPassword:     kingpin.Flag("admin-password", "Hashed admin password").String(),
+		AdminPasswordFile: kingpin.Flag("admin-password-file", "Hashed admin password from file").String(),
```
Can you update the flag description to `Path to the file containing the password for the admin user`

deviantony commented on api/cmd/portainer/main.go:
```diff
@@ -212,12 +214,28 @@ func main() {
+		if err != nil {
+			log.Fatal(err)
+		}
+		// As the file content is in plain text we need to hash it
```
You can remove the comment.

deviantony commented:
Hey @nenadilic84, have you tested this? I was unable to login using this new flag either by using a secret or by mounting my password file into the container... I've tried to figure out where the error comes from but I don't know.

Seems that the hash generated from the file is different from a hash using the string directly. I've added some debug and here is the output:

```
2017/09/21 14:26:06 Password from file: admin123
2017/09/21 14:26:07 Hash from file: $2a$10$xvhfVR4h9yGv6JIiy9rt2.ib2EzetnnfJMEalr0hQmvn2ZMPyKoNW
2017/09/21 14:26:07 Hash from string admin123: $2a$10$D0mql./ZnlI/qeuBejRcw.hbJGZeqUzpXsOQLNmrSQYks9hQWtMOq
``` 

Using the following debug instructions:

```golang
content, err := ioutil.ReadFile(*flags.AdminPasswordFile)
if err != nil {
	log.Fatal(err)
}
log.Printf("Password from file: %v", string(content))
AdminPasswordHash, err = cryptoService.Hash(string(content))
log.Printf("Hash from file: %v", AdminPasswordHash)
AdminPasswordHash, err = cryptoService.Hash("admin123")
log.Printf("Hash from string admin123: %v", AdminPasswordHash)
```

deviantony commented on api/cmd/portainer/main.go:
```diff
@@ -212,12 +214,27 @@ func main() {
 
-	if *flags.AdminPassword != "" {
-		log.Printf("Creating admin user with password hash %s", *flags.AdminPassword)
+	AdminPasswordHash := ""
```
The `AdminPasswordHash` variable should be `adminPasswordHash`

nenadilic84 replied:
Yup, this makes sense

deviantony commented on api/cmd/portainer/main.go:
```diff
@@ -212,12 +214,27 @@ func main() {
-		log.Printf("Creating admin user with password hash %s", *flags.AdminPassword)
+	AdminPasswordHash := ""
+	if *flags.AdminPasswordFile != "" {
+		content, err := ioutil.ReadFile(*flags.AdminPasswordFile)
```
I'd rather see this in the `file` package rather to have an import to `ioutil` in this file.

nenadilic84 replied:
The `file` package has a `Service Struct` that operates on files that are relative to the `/data` path, and in case of passing secrets the files are provided with the full path which is usually not in the `/data`. If I move this to the file I would then create a global function on getting the string from a file only... Let me know if that is what you had in mind?

deviantony replied:
Yeap, include it as a global generic function that returns a string from a file.

nenadilic84 commented:
Didn't test is end to end. Only with the string... And is quite strange that this is happening.
Need to take a closer look.

Thanks for taking time to verify

nenadilic84 commented:
Hi @deviantony, I did an end to end test and seems that all is working fine.
The reason why you see different hashes is that of the bcrypt algorithm. 
You should do the comparison by using the function `cryptoService.CompareHashAndData(adminPasswordHash, "admin123")`

deviantony commented:
Woops, you're right, sorry for that.

@nenadilic84 could you send me the steps that you used to test it?

nenadilic84 commented:
So the steps are quite straightforward:
```
./build.sh test
docker tag portainer/portainer:linux-amd64 nenadilic84/portainer:test
echo -n "test1234" > portainer-pass
cp portainer-pass /tmp/portainer/
# run it
docker run -it \
    -p 9000:9000 \
    -v /var/run/docker.sock:/var/run/docker.sock \
    -v /tmp/portainer:/data \
    --name portainer \
    portainer/portainer:linux-amd64 --admin-password-file '/data/portainer-pass'

# or in docker swarm
echo -n test1234 | docker secret create portainer-pass -
docker service create \
    --name portainer \
    --secret portainer-pass \
    --publish 9000:9000 \
    --constraint 'node.role == manager' \
    --mount type=bind,src=/var/run/docker.sock,dst=/var/run/docker.sock \
    nenadilic84/portainer:test \
    --admin-password-file '/run/secrets/portainer-pass' \
    -H unix:///var/run/docker.sock
```
While writing down the steps and comparing the the comment above I realized you don't have the `echo -n` but just `echo` which adds the newline character `0a` to the file/secret.

Try it out with:
```
echo $PORTAINER_PASS | docker secret create portainer_password -
```

I can also add, while reading a file, to ignore `newline` character, but not sure if we want to do that... Let me know on how to proceed...

nenadilic84 commented:
Anyway, I've updated the other things you've mentioned and added trimming the newline in the `GetStringFromFile` function.
If you think that it's not necessary I can change the line: `return strings.TrimSuffix(string(content), "\n"), nil` to something like: `return string(content), nil`

deviantony commented on api/file/file.go:
```diff
@@ -162,3 +165,13 @@ func (service *Service) createFileInStore(filePath string, r io.Reader) error {
+		return "", err
+	}
+
+	return strings.TrimSuffix(string(content), "\n"), nil
```
I agree with your last point, we should not remove the newline character.

nenadilic84 commented:
Updated the PR.
@deviantony, let me know if it's all good.

deviantony commented:
Thanks @nenadilic84 :+1:
